Enable omero-build to use this repository

joshmoore opened the pull request:
In an effort to allow the decoupled submodule repositories
to also be able to use server integration tests in travis,
the relationship between omero-build and this repository
will need to be changed. For example, a PR in omero-server
will need a matching PR in omero-build which will build
all the component jars and then inject them into the Dockerfile
in this repository. This is done by:

 (1) appending the necessary versions to etc/omero.properties:

```
    echo versions.omero-blitz >> etc/omero.properties
```

 (2) using the omero-build image as the base image

```
    --build-arg=BUILD_IMAGE=...name of omero-build image...
```

joshmoore committed:
Enable omero-build to use this repository

In an effort to allow the decoupled submodule repositories
to also be able to use server integration tests in travis,
the relationship between omero-build and this repository
will need to be changed. For example, a PR in omero-server
will need a matching PR in omero-build which will build
all the component jars and then inject them into the Dockerfile
in this repository. This is done by:

 (1) appending the necessary versions to etc/omero.properties:

```
    echo versions.omero-blitz >> etc/omero.properties
```

 (2) using the omero-build image as the base image

```
    --build-arg=BUILD_IMAGE=...name of omero-build image...
```

joshmoore commented:
@sbesson : I could use your suggestion here. When build with M3 jars, `ome-resolver` fails as on travis in https://travis-ci.org/openmicroscopy/openmicroscopy/builds/500214378 since only user-maven is attempted:

```
	==== user-maven: tried
	  /home/build/.m2/repository/org/openmicroscopy/omero-blitz/5.5.0-m3/omero-blitz-5.5.0-m3-python.zip
``` 

which is why I previously hard-coded to `ome-artifactory`. But for the "use .m2 for SNAPSHOTS", I want *only* `user-maven`. I can turn this into "Yet Another Ivy Property" but that seems overkill. Do you know of an alternative?

jburel commented:
Without the challenge to ``ivysettings``, it is currently not possible to build locally
 with snapshot.
So we may want to extract that depending on the timeline

joshmoore commented:
@jburel : sorry, I don't understand. What are you trying to build? If you're saying that this PR isn't working, I know. That's what I'm trying to figure out.

jburel commented:
Sorry I am testing the "m4" PRs from @rgozim locally so I am using the artifacts generated locally

joshmoore committed:
Fix ivysettings by unifying classifiers

joshmoore commented:
Couple of additional thoughts:

 * the ivy/pom lines might need `(-[classifier])` but it's unclear if that will work
 * this may also help with "every other build fails" issues
 * but it's failing since ivy seems to only find `user-maven` in the `5.5.0-m3` case

joshmoore committed:
Various semi-fixes to the artifactory issue that all work together!

The base problem as outlined on gh-5968 is that release artifacts
(e.g. 5.5.0-m3) can be used with `ome-artifactory` resolver while
local snapshot artifacts (e.g. 5.5.0-SNAPSHOT) can be found with
the `user-maven` resolver, but both don't work together. Fixes &
their relative usefulness:

 * classifier everywhere: didn't help, but should
 * returnFirst: think it helps, possible performance impact
 * all tools depend on python.zip: finally working, but ugly

joshmoore commented:
cc: @jburel and @sbesson (see commit message)

jburel commented:
With the ultimate goal in mind (getting rid of ivy), the (a bit ugly) solution  that allows everything to work  together seems acceptable

sbesson commented:
Tested locally a build with the last commit from a fresh `~/.m2` repository and I can confirm the compilation passes.

Having parallelly tried to look into Ivy artifact resolution logic, I suspect most of our problems arises from the nature of the `omero-blitz` artifact which escapes the classical `POM + JAR (+ Javadoc + sources)` composition with the existence of an additional `python` artifact of type `zip`.

https://github.com/openmicroscopy/openmicroscopy/pull/5968/commits/a44ffe6108d2be97ed8409efd8ad098895c0e8ed#diff-f17d700ee90f1db676bcc7f85037fb3eR36 looks like a way to tell Ivy that `omero-blitz` is more than a JAR the first time the artifact is retrieved even if the `python` is unnecessary for this component. We might want to note this.

joshmoore commented:
Note: current travis failure is going to require a merge on omero-test-infra.

joshmoore committed:
Run all integration tests by default

joshmoore committed:
Temporarily pin to srv-compose branch

joshmoore committed:
Retry SSL fix

joshmoore committed:
Used merged omero-test-infra

joshmoore committed:
Re-disable ssl 'fix'

joshmoore committed:
Configure travis-build to run unit tests

joshmoore committed:
Install flake8 and pytest in dockerfile

joshmoore committed:
Also build java for travis

sbesson commented on Dockerfile:
```diff
@@ -22,17 +22,18 @@ ARG RUN_IMAGE=openmicroscopy/omero-${COMPONENT}:latest
 # openjdk:8 is "stretch" or Debian 9
 RUN pip install https://github.com/ome/zeroc-ice-py-debian9/releases/download/0.1.0/zeroc_ice-3.6.4-cp27-cp27mu-linux_x86_64.whl
-RUN adduser omero
+RUN id 1000 || useradd -u 1000 -ms /bin/bash build
```
Nice way to conditionally add the user and enable the base image to be overriden at build time.

sbesson commented on components/tools/travis-build:
```diff
@@ -123,18 +123,7 @@ build_cpp()
-        ( cd components/tools/OmeroPy/;
-          python setup.py test -t $TEST -m "not broken"
-        )
-    done
```
Is this covered in the `test-integration` target or something we want to come back to later ?

joshmoore replied:
`test-integration` is a superset of this garbage. Note also that `integration` was only ever used from omero-test-infra --> `travis-build`. The trade-off is that just building touched tests built _fast_ but not every PR touched the tests.

sbesson reviewed (commented):
Excluding the diff of the changes already merged in #5971, the changes in `Dockerfile` and `components/travis` look fine as a step to couple this to `omero-build` and other upstream repositories.

Travis is green so I am fine with getting it merged as starting to consume it actively.

joshmoore commented:
Ok. After an extended call, merging. What this *doesn't* yet deal with:

 * how to include an ant build into the gradle-based omero-build repo
 * how to use version numbers rather than `SNAPSHOT` for pushing to nexus
 * how to verify PRs against this repository independently of `omero-build`

cc: @rgozim @manics @mtbc @jburel @sbesson
